README.md
# speedometer

Customizable speedometer widget for Flutter (Android, iOS) with onComplete observable method.

![Image](https://github.com/AlxGration/FlutterWidget/blob/main/img/screen.gif)

## Getting started

In order to use this package, add the following dependency in your Flutter project.
```yaml
dependencies:
  alxgration_speedometer: ^0.0.1
```
and then press "Pub get" if you are using AndroidStudio or run `flutter packages upgrade` otherwise.

In your Dart code, add import:
```dart
import 'package:alxgration_speedometer/speedometer.dart';
```

## Usage

Short example for package use. Longer examples in `/example` folder. 

```dart
Speedometer(
    size:200,
    minValue: 0,
    maxValue: 10,
    currentValue: progress,
    barColor: Colors.purple,
    pointerColor: Colors.black,
    displayText: "km/h",
    displayTextStyle: TextStyle(fontSize: 14, color: Colors.deepOrange),
    displayNumericStyle: TextStyle(fontSize: 24, color: Colors.red),
    onComplete: (){
        print("ON COMPLETE");
    },
);
```
## Package
You can find the package [here](https://pub.dev/packages/alxgration_speedometer)

CHANGELOG.md
## 0.0.2
* Speedometer: fix linter issues to get higher score.

## 0.0.1
* Speedometer: initial release.


test/speedometer_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:alxgration_speedometer/speedometer.dart';

void main() {
  test('check creation of widget', () {
    final speedometer = Speedometer();
  });
}


lib/ViewPainter.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';

class ViewPainter extends CustomPainter {
  ViewPainter(
      {this.startAngle = 0, this.sweepAngle = 0, this.color = Colors.white});

  final double startAngle;
  final double sweepAngle;
  final Color color;

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTRB(size.width * 0.05, size.width * 0.05,
        size.width * 0.95, size.height * 0.95);
    final startAngle = math.pi / 12 * this.startAngle;
    final sweepAngle = math.pi / 12 * this.sweepAngle;
    final useCenter = false;
    final paint = Paint()
      ..color = this.color
      ..style = PaintingStyle.stroke
      ..strokeWidth = size.width * 0.1;
    canvas.drawArc(rect, startAngle, sweepAngle, useCenter, paint);
  }

  @override
  bool shouldRepaint(CustomPainter old) {
    return false;
  }
}

//cuts unused angle of ring
class TriangleClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    final path = Path();
    path.moveTo(size.width * 0.5, 0.0);
    path.lineTo(size.width, size.height);
    path.lineTo(0.0, size.height);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(TriangleClipper oldClipper) => false;
}

//arrow of pointer
class PointerClipper extends CustomClipper<Path> {
  @override
  Path getClip(Size size) {
    final path = Path();
    path.moveTo(size.width * 0.5, size.width * 0.5);
    path.lineTo(size.width * 0.5 + size.width / 30, size.height * 0.5);
    path.lineTo(size.width * 0.5 + 1, size.height - size.width / 30);
    path.lineTo(size.width * 0.5 - 1, size.height - size.width / 30);
    path.lineTo(size.width * 0.5 - size.width / 30, size.height * 0.5);
    path.close();
    return path;
  }

  @override
  bool shouldReclip(PointerClipper oldClipper) => false;
}


lib/speedometer.dart
library alxgration_speedometer;

import 'dart:math' as math;
import 'package:flutter/material.dart';
import 'ViewPainter.dart';

class Speedometer extends StatefulWidget {
  static const TextStyle defaultNumericStyle =
      TextStyle(color: Colors.black, fontSize: 24);
  static const TextStyle defaultTextStyle =
      TextStyle(color: Colors.black, fontSize: 18);

  Speedometer(
      {this.size = 200,
      this.minValue = 0,
      this.maxValue = 100,
      this.currentValue = 0,
      this.barColor = Colors.purple,
      this.pointerColor = Colors.black,
      this.displayNumericStyle = defaultNumericStyle,
      this.displayText = "",
      this.displayTextStyle = defaultTextStyle,
      this.onComplete});

  final double size;
  final int minValue;
  final int maxValue;
  final int currentValue;
  final Color barColor;
  final Color pointerColor;
  final TextStyle displayNumericStyle;
  final String displayText;
  final TextStyle displayTextStyle;
  final VoidCallback? onComplete;

  @override
  _SpeedometerState createState() => _SpeedometerState();
}

class _SpeedometerState extends State<Speedometer> {
  @override
  Widget build(BuildContext context) {
    double _size = widget.size;
    int _minValue = widget.minValue;
    int _maxValue = widget.maxValue;
    int _currentValue = widget.currentValue;
    double startAngle = 3.0;
    double endAngle = 21.0;

    //setting pointer(speedometer) value
    double _pointerAngle = 0;
    if (_minValue <= _currentValue && _currentValue <= _maxValue) {
      _pointerAngle = (((_currentValue - _minValue) * (endAngle - startAngle)) /
              (_maxValue - _minValue)) +
          startAngle;
    } else if (_currentValue < _minValue) {
      _pointerAngle = startAngle;
    } else if (_currentValue > _maxValue) {
      _pointerAngle = endAngle;
    }

    // call listener.onComplete in case pointer achieved the end.
    if (_currentValue >= _maxValue) {
      if (widget.onComplete != null) {
        widget.onComplete!();
      }
    }

    return Container(
      color: Colors.transparent,
      child: Center(
        child: Directionality(
          textDirection: TextDirection.ltr,
          child: Container(
            width: _size,
            height: _size,
            child: Stack(
              children: <Widget>[
                Container(
                  alignment: Alignment.center,
                  padding: EdgeInsets.all(_size * 0.075),
                  child: CustomPaint(
                    /* speedometer circle drawing */
                    size: Size(_size, _size),
                    painter: ViewPainter(
                        startAngle: 9, sweepAngle: 18, color: widget.barColor),
                  ),
                ),
                Container(
                  //triangle among center of pointer and bottom right-left box of widget
                  alignment: Alignment.bottomCenter,
                  child: ClipPath(
                    clipper: TriangleClipper(),
                    child: Container(
                      width: _size,
                      height: _size * 0.5,
                      color: Colors.transparent,
                    ),
                  ),
                ),
                Container(
                  // center circle of pointer
                  alignment: Alignment.center,
                  child: Container(
                    width: _size * 0.1,
                    height: _size * 0.1,
                    alignment: Alignment.center,
                    decoration: new BoxDecoration(
                      color: widget.pointerColor, //
                      shape: BoxShape.circle,
                    ),
                  ),
                ),
                Container(
                  // pointer (triangle) itself
                  alignment: Alignment.center,
                  child: Transform.rotate(
                    angle: math.pi / 12 * _pointerAngle,
                    child: ClipPath(
                      clipper: PointerClipper(),
                      child: Container(
                        width: _size * 0.9,
                        height: _size * 0.9,
                        color: widget.pointerColor,
                      ),
                    ),
                  ),
                ),
                Container(
                  // value text displayed below speedometer
                  alignment: Alignment.bottomCenter,
                  child: Padding(
                    padding: EdgeInsets.fromLTRB(0, 0, 0, _size * 0.1),
                    child: Text(
                      widget.currentValue.toString(),
                      style: widget.displayNumericStyle,
                    ),
                  ),
                ),
                Container(
                  // measurement text displayed below speedometer
                  alignment: Alignment.bottomCenter,
                  child: Padding(
                    padding: EdgeInsets.fromLTRB(0, 0, 0, 0),
                    child: Text(
                      widget.displayText,
                      style: widget.displayTextStyle,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:alxgration_speedometer/speedometer.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
          colorScheme: ColorScheme.fromSwatch(
              primarySwatch: Colors.deepPurple,
              accentColor: Colors.deepOrangeAccent)),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int progress = 0;
  int maxVal = 5;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("My Speedometer Widget"),
        ),
        body: Column(children: [
          Speedometer(
            size: 200,
            minValue: 0,
            maxValue: maxVal,
            currentValue: progress,
            barColor: Colors.purple,
            pointerColor: Colors.black,
            displayText: "km/h",
            displayTextStyle: TextStyle(fontSize: 24, color: Colors.deepOrange),
            displayNumericStyle: TextStyle(fontSize: 24, color: Colors.red),
            onComplete: () {
              showSnackbar();
            },
          ),
          ElevatedButton(
              onPressed: () {
                increaseProgress();
              },
              child: Text("start"))
        ])
    );
  }



  void showSnackbar() {
    WidgetsBinding.instance.addPostFrameCallback((_){
      ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
            content: Text("ON COMPLETE"),
          )
      );
    });
  }

  void increaseProgress() async {
    for (int i = 0; i < maxVal; i++) {
      var j = Future.delayed(const Duration(seconds: 1), () {
        return i;
      });
      progress = (await j) + 1;
      setState(() {});
    }
  }
}
